### ISSUE TITLE
Add syntax extension to add warning blocks

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/rust#79710]
Following https://github.com/rust-lang/rust/pull/79677, we realized that rustdoc users would prefer to have a syntax extension rather than using plain HTML to have warning blocks.

@dtolnay suggested the following:

```text
/// WARNING: it doesn't work yet.
```

@Darksonn suggested:

```text
[[info]]
| Many of Tokio's types are named the same as their synchronous equivalent in
| the Rust standard library. When it makes sense, Tokio exposes the same APIs
| as `std` but using `async fn`.
```

What do you think?

cc @rust-lang/rustdoc

**Comment by CraftSpider** [comment]
Personal Opinions:
The first style will look fine for short forms, but I'm not sure for longer text. A possible syntax:
```
/// WARNING: This is a lot of text, continued
///          on multiple lines, and what if I want more formatting, or another block?
///          *this is interesting*
```
Which to me pushes the text too far right. Not having any spacing, but just separating with newlines, seems hard to parse at a glance what is part of the WARNING, and prevents multiple line breaks inside the warning. Also, it's not as extensible. INFO can be added, but if we ever want more extensions, they would increasingly possibly conflict with other things.
The second style makes it immediately clear what is part of the block, and is unlikely to accidentally conflict if we ever want to add more block styles. It also supports multiple lines or inner blocks trivially. A final advantage is that it may support nested blocks, in case those ever come to be useful.
Overall: I prefer the second form, both aesthetically and for future-proofing

**Comment by camelid** [comment]
I think I agree with `@CraftSpider` in preferring the second form. I like the first form's simplicity, but it also seems surprising that `WARNING:` is treated as a 'keyword' and will change the formatting of the docs.

**Comment by GuillaumeGomez** [comment]
That's why we have this debate. :stuck_out_tongue:

**Comment by Manishearth** [comment]
I _personally_ actually prefer stabilizing a CSS class, I think we should do more of that, just because parsing can get annoying. But this is not that hard to parse anyway.

The problem with `WARNING:` is precisely that you can't scope it easily: what if you want to add code blocks? bullet points?

**Comment by jyn514** [comment]
> I personally actually prefer stabilizing a CSS class

The problem with this is it ties rustdoc to the HTML backend; like I said [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/warning.20sections.20in.20rustdoc.3A.20.2379677/near/218895331) nothing rustdoc currently does guarentees that it will generate HTML from the markdown. Changing that means that to make the JSON backend have the same info, we'd have to add a full HTML parser to rustdoc just to read the HTML in the *markdown*, because pulldown does not parse HTML for us. I really think we should avoid making the JSON backend a second-class citizen.

**Comment by jyn514** [comment]
Also there's not much point to this if it's just a little CSS IMO, you can add that yourself with `--html-after-content` and it will look just as good and be more flexible since you can control the styles yourself.

**Comment by jyn514** [comment]
I'm not a giant fan of the `[[info]]` syntax because it looks odd when rendered by anything other than rustdoc, which I'd rather avoid, one of the major design concerns for intra-doc links was compatibility with other markdown parsers: https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md#standard-conforming-markdown

**Comment by Manishearth** [comment]
> The problem with this is it ties rustdoc to the HTML backend

So, I agree, but I'm not convinced that the warning class _is_ something that needs to be surfaced in the JSON backend as anything special.

Also, bear in mind, HTML is a part of markdown: valid HTML (well a subset of HTML) is valid markdown. Users already can `<div style=foo>`, and they do that in many cases.

Markdown parsers already need to be able to handle HTML. There's nothing new there. Whatever consumes the JSON backend will need to be able to handle that too.


> Also there's not much point to this if it's just a little CSS IMO, you can add that yourself with `--html-after-content` and it will look just as good and be more flexible since you can control the styles yourself.

I mean, yes, in theory, in practice people don't do this because of the barrier to entry (you have to get your tools to use that, including docs.rs)

**Comment by Manishearth** [comment]
> because pulldown does not parse HTML for us

Also, this is incorrect, pulldown absolutely parses HTML. It has to, like I said valid HTML is valid Markdown. It doesn't parse it _thoroughly_, I believe you can mess it up with sufficiently complex HTML, but it parses HTML.

-----

What this changes is subtler: up till now you did not need any particular stylesheet to acceptably render rustdoc markdown from the JSON backend on your own (just use your favorite markdown renderer's stylesheet), but now you will need other css classes to handle rendering warnings as warnings.

**Comment by dtolnay** [comment]
> The problem with `WARNING:` is precisely that you can't scope it easily: what if you want to add code blocks? bullet points?

I do not agree that this is a problem. I've used Phabricator's implementation of this approach happily and it is by design that warning banners are one paragraph. It is not the intention that you would put entire sections of documentation into a warning banner. They are for a brief callout. It's fine to refer to a subsequent *section* of the documentation *from* the callout if needed.

I suppose this is a matter of taste but it's one where I believe the opinionatedness of the syntax is a positive thing.

**Comment by Manishearth** [comment]
So I discussed this a bit with @jyn514 on the side, we agree that the effect on the JSON backend isn't major since HTML is _already_ a subset of markdown and docs use HTML already (especially for images and tables).

My position is as follows:

 - I strictly prefer HTML over some other scoped syntax with nuances people have to learn.
 - I don't think there's as much value in unscoped syntax

which makes my position "HTML or nothing", basically. I'm not wholly against unscoped single-line syntax. We `NOTE:` and `EXAMPLE:` syntax in W3C specs and you invariably need to add more stuff to it (there's an HTML class in these specs for this reason).

Furthermore, Rust has a strong culture of wrapping text, so this means we need `WARNING:` to work well while wrapped, and that means we're again introducing some form of scoping mechanism (e.g. indenting the text), which _editors_ will have to learn, as well.

**Comment by masklinn** [comment]
I don't know about admonitions specifically but there seems to be something of a community habit to use "magic tags" on fenced code blocks for this sort of features (e.g. github has a `suggestion` tag), maybe that could be an option? I don't know if pulldown_cmark has a hook for manipulating code blocks but…

So a warning would be

    ```warning
    as much warning text as you want
    ```

or maybe the tag would be `admonition:<type>` to leave room for things like note/todo/danger/…

Alternatively, take a gander at how other markdown implementations extend the language for this sort of features? e.g. [python-markdown](https://python-markdown.github.io/extensions/admonition/) seems to be using `!!!<tag>` for admonitions.

**Comment by camelid** [comment]
> I don't know about admonitions specifically but there seems to be something of a community habit to use "magic tags" on fenced code blocks for this sort of features (e.g. github has a `suggestion` tag), maybe that could be an option? I don't know if pulldown_cmark has a hook for manipulating code blocks but…

Hmm, but `suggestion` is still for code, even if it has some extra behavior.

Aside from the design concerns, this would be feasible to implement: we just extract the language string from the code block with pulldown.

**Comment by gideongrinberg** [comment]
Could you use something that's already built into markdown? For example, rustdoc could take `> :warning: Some warning here` and render it as an admonition.

**Comment by GuillaumeGomez** [comment]
That's what https://github.com/rust-lang/rust/issues/79710#issuecomment-750350933 suggested: to use markdown and make some changes on rustdoc side. I think the codeblock approach is the best personally.

**Comment by gideongrinberg** [comment]
> That's what [#79710 (comment)](https://github.com/rust-lang/rust/issues/79710#issuecomment-750350933) suggested: to use markdown and make some changes on rustdoc side. I think the codeblock approach is the best personally.

My bad, I didn't see that.

**Comment by jyn514** [comment]
I wonder if it makes more sense to add this as a proc-macro. Rustdoc can still provide the CSS classes and things, and you could use them with HTML directly, but you could use `warning!("warning text here")` or something and it would expand to the correct HTML tags and classes.

**Comment by GuillaumeGomez** [comment]
Might be better to have a `doc(warning = "...")` then, no?

**Comment by Manishearth** [comment]
Is that going to be a special header similar to the "unsafe"/"experimental"/etc ones, or is that going to also be used within docs in an order-sensitive way? If the latter we should probably just use a CSS class; mixing `///` and `#[doc]` is ugly

**Comment by GuillaumeGomez** [comment]
Agreed. Among the suggested ideas, my favourite one remains the code block with "warning" as tag:

    ``` warning
    some text
    ```

**Comment by Manishearth** [comment]
Yeah, but triple backticks is also for preformatted text without further markup, so you won't be able to e.g. use text formatting (or other code) in there

**Comment by GuillaumeGomez** [comment]
True... Finding the right syntax is much more complicated than expected...

**Comment by masklinn** [comment]
> True... Finding the right syntax is much more complicated than expected...

That's the issue of using markdown, for all that it has a pretty nice baseline syntax it really was not designed for extensibility.

**Comment by GuillaumeGomez** [comment]
We can always use `<warning>`: it'd allow to have text formatting inside it.

**Comment by Manishearth** [comment]
I would use `div class=warning`, it's not quite kosher to invent arbitrary HTML tags (though you can use Custom Elements, which is more heavyweight)

**Comment by GuillaumeGomez** [comment]
Sounds good to me. It's funny because we kinda came back to the [starting point](https://github.com/rust-lang/rust/pull/79677). :rofl:

**Comment by GuillaumeGomez** [comment]
I just thought: we could use `<warning>` tag in the markdown and then transform it into `<div class="warning">` in the generated HTML?

**Comment by jyn514** [comment]
My main objection to *all* of these is that rustdoc shouldn't be adding custom markdown syntax that no one else supports. Any new syntax we use should look reasonable even if rendered with the default `pulldown-cmark` renderer.

**Comment by GuillaumeGomez** [comment]
Well, it remains valid markdown since it's HTML (even though it's custom tag). I'm completely fine with `<div class="warning">`, but it might be simpler/better for users to simply write `<warning>`.

**Comment by GuillaumeGomez** [comment]
Coming back to this (because of #106453), it seems like the conversation kinda agreed on adding a CSS class directly and then it would look like this in the docs: `<div class="warning">`. I'll send a PR in the next days and then we can open an FCP to see if everyone is ok with it (and if not, we'll have code to help the debate).

**Comment by ogoffart** [comment]
> **Warning**
> Did you know that GitHub markdown has an extension for Notes and Warnings:
> https://github.com/community/community/discussions/16925

> **Note**
> Maybe it may make sense to try to follow these existing extension. 
> I once wanted to do that in our README.md which is `#![doc = include_str!("README.md")]`

**Comment by coastalwhite** [comment]
@ogoffart This indeed seems like a better solution than the HTML block. It also plays better with something @jyn514 said about not linking it too much too HTML.

**Comment by GuillaumeGomez** [comment]
They also seemed to not be too favourable to anything that needs some extra parsing from rustdoc iuc. I'm fine with bot the github solution and the CSS class.

**Comment by coastalwhite** [comment]
Maybe one more comment to make here. The [Obsidian.md](https://obsidian.md/) note-taking app, which uses Markdown also has a similar syntax to create these blocks. It is called [Callouts or admonitions](https://help.obsidian.md/How+to/Use+callouts). It uses a syntax similar to the syntax that is used for doc intra-links currently, and it is fully compatible with them.

```markdown
> [!info]
> This is an information block
```

```markdown
> [!warning]
> This is an warning block
```

I am uncertain if this is any better than what @ogoffart suggested from the GitHub syntax, but because of the similarity to the doc intra-link syntax. I do feel like this might be interesting because it should go well with backwards compatibility. However, this may look worse than the GitHub syntax solution when not going through cargo doc.

**Comment by Manishearth** [comment]
I thought I'd already left this comment, but I'd really like to avoid adding more markdown _syntax_ where possible. This feature works _fine_ without special syntax, I don't think we need it.

**Comment by coastalwhite** [comment]
@Manishearth Although, I might not agree fully, I can fully appreciate and understand the reasoning you left in the corresponding PR.

Still, I do think it might be worth to standardize some kind of syntax for some kind of warning block. This does not actually require customized markdown. Alternatively, I can suggest adding the GitHub flavor with emojis to the [Rust API Guidelines documentation section](https://rust-lang.github.io/api-guidelines/documentation.html). This may help in standardizing short form warnings and notes within documentation.

I do value people's opinion on this. Is this a possible compromise?

**Comment by Manishearth** [comment]
That's a valid alternative, I think the yellow color is pretty important though, and it's useful for _themes_ to be able to make styling decisions about warnings so I do think that having _some_ way of marking it is important.

I just don't think the way of marking it should be some magic invocation as opposed to CSS classes.

(And I think that the bar for a feature that is adding new semantics to the Rustdoc markdown parser is much, _much_ higher than that of one stabilizing a CSS class)

**Comment by coastalwhite** [comment]
100% agree. I will look at the Rust API guidelines issue and possibly an RFC to rust to track the adoption in GitHub and possibly eventually merge.

**Comment by GuillaumeGomez** [comment]
This is a good point. Once we have something merged here, I'll send a PR there to have an equivalent.

**Comment by ehuss** [comment]
I'd be happy to add some CSS rules to mdbook if that is what you are proposing.

**Comment by toastal** [comment]
FYI, using a `<blockquote>` is _not_ semantic and Microsoft GitHub is setting a bad precedence.

> A block quote marker, optionally preceded by up to three spaces of
> indentation, consists of (a) the character `>` together with a
> following space of indentation, or (b) a single character `>` not
> followed by a space of indentation.

— CommonMark Spec v0.30, https://spec.commonmark.org/0.30/#block-quotes

`>` denotes a `<blockquote>` in Markdown and is rendered as such.

> The blockquote element represents a section that is quoted from
> another source.

— W3C HTML spec, https://html.spec.whatwg.org/multipage/grouping-content.html#the-blockquote-element

Unless you are quoting a source, this is not the correct element to use. While I disagree with even using `>` altogether, at least some tools out there are at least translating the blockquote in admonition context into a `<div>` (div has no semantic meaning which is unfortunately the best HTML option for admonitions). It would be ill-advised to follow Microsoft GitHub’s Markdown syntax fork&mdash;especially with their disregard for semantics. Similarly using a `<pre>` block would also be breaking semantics as an admonition is not preformatted text. 


CommonMark _does_ have [an RFC for generic directives](https://talk.commonmark.org/t/generic-directives-plugins-syntax/444) that could be used if the `<div class="warning">` solution was preferred. There’s also other reasonable options noted here that don’t break semantics.

**Comment by GuillaumeGomez** [comment]
Using `<div class="warning">` is what we agreed upon in https://github.com/rust-lang/rust/pull/106561/files#diff-625c2b4568f0a7b1c8233bee4e602cdd31b6d88bc06d82dbe33590bf213cd41dR265. The debate now is on the look itself.

**Comment by jhpratt** [comment]
Unfortunately `.warning` isn't great because markdown has a "boundary" when HTML tags are used. I wanted to add a warning indicating that users probably wanted another trait, but intra-doc links don't work (which is the correct behavior per markdown spec).

**Comment by GuillaumeGomez** [comment]
Works perfectly fine though (well, minus the UI bug I just discovered haha):

![image](https://github.com/rust-lang/rust/assets/3050060/f127bc40-a65d-4ca7-b3cd-38be5226d182)

I used this code:

```rust
/// <div class="warning">
///
/// [bar]
///
/// </div>
pub fn foo() {}

pub fn bar() {}
```

I'll send a PR to improve documentation.

**Comment by jhpratt** [comment]
Ah. Well I suppose the issue wasn't with intra-doc links, but with inline code blocks. What I tried was that but with ``[`bar`]``, and it failed.

**Comment by GuillaumeGomez** [comment]
Works too:

![image](https://github.com/rust-lang/rust/assets/3050060/5314bfc2-000c-40f7-a41b-c09329637c6b)

**Comment by GuillaumeGomez** [comment]
I opened https://github.com/rust-lang/rust/pull/119245 which adds extra explanation on how to use markdown inside html tags. That should solve your issue.

**Comment by jhpratt** [comment]
That note is exactly what I was just looking in to. I had it inline, given my familiarity with HTML. TIL that matters.

**Comment by GuillaumeGomez** [comment]
It doesn't hurt to have it and if it can helps users, well, why not doing it. :)
